File: src/main/java/com/example/usercrud/Presentation/UserController.java
package com.example.usercrud.Presentation;

import com.example.usercrud.Business.Entity.ErrorResponse;
import com.example.usercrud.Business.Entity.User;
import com.example.usercrud.Business.Service.UserService;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.client.RestTemplate;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.util.*;

@RestController
@RequiredArgsConstructor
public class UserController {
    final UserService userService;
    ErrorResponse errorResponse;

    public static boolean isValidEmailAddress(String email) {
        String ePattern = "^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@((\\[[0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3}\\.[0-9]{1,3}\\])|(([a-zA-Z\\-0-9]+\\.)+[a-zA-Z]{2,}))$";
        java.util.regex.Pattern p = java.util.regex.Pattern.compile(ePattern);
        java.util.regex.Matcher m = p.matcher(email);
        return m.matches();
    }

    @GetMapping("/lol")
    public void kek(HttpServletRequest request) throws JsonProcessingException {
        String userIP = "84.240.214.26";
        System.out.println(userIP);


        RestTemplate restTemplate = new RestTemplate();
        String result = restTemplate.getForObject("http://ip-api.com/json/" + userIP + "?fields=status,country", String.class);

        ObjectMapper mapper = new ObjectMapper();
        Map map = mapper.readValue(result, Map.class);
        System.out.println((String) map.get("country"));

    }

    @PostMapping("/add")
    public ResponseEntity addUser(@RequestBody User user, HttpServletRequest request) throws JsonProcessingException {
        errorResponse = new ErrorResponse(new Date());
        if(user == null){
            errorResponse.getErrorDescriptions().add("No user info");
            return ResponseEntity.badRequest().body(errorResponse);
        }
        if(user.getFirstName() == null || user.getFirstName().isEmpty()){
            errorResponse.getErrorDescriptions().add("First name is empty");
        }

        if(user.getEmail() == null){
            errorResponse.getErrorDescriptions().add("No email");
        } else if (!isValidEmailAddress(user.getEmail())){
            errorResponse.getErrorDescriptions().add("Invalid email");
        }

        if(userService.existsByEmail(user.getEmail())){
            errorResponse.getErrorDescriptions().add("User with this email already exists");
        }
        if(user.getGender() == null){
            errorResponse.getErrorDescriptions().add("No gender");
        }
        if(!errorResponse.getErrorDescriptions().isEmpty()){
            return ResponseEntity.badRequest().body(errorResponse);
        }
        String userIP = request.getRemoteAddr();
        System.out.println(userIP);

        // if we run app on localhost getRemoteAddr() always returns smth like localhost;127.0.0.1;0:0:0:0:0:0:0:1
        if(Objects.equals(userIP, "0:0:0:0:0:0:0:1") || Objects.equals(userIP, "127.0.0.1")){
            System.out.println(userIP);
            user.setCountry("Kazakhstan");
        } else {
            RestTemplate restTemplate = new RestTemplate();
            String result = restTemplate.getForObject("http://ip-api.com/json/" + userIP + "?fields=status,country", String.class);

            ObjectMapper mapper = new ObjectMapper();
            Map map = mapper.readValue(result, Map.class);

            if(map.get("status") == "fail"){
                user.setCountry("undefined");
            } else {
                user.setCountry((String) map.get("country"));
            }
        }
        return ResponseEntity.ok(userService.addUser(user));
    }

    @DeleteMapping("/id/{id}")
    public ResponseEntity deleteUserById(@PathVariable Long id){
        errorResponse = new ErrorResponse(new Date());
        Optional<User> user = userService.deleteById(id);
        if(user.isEmpty()){
            errorResponse.getErrorDescriptions().add("User with this id doesn't exist");
            return ResponseEntity.badRequest().body(errorResponse);
        }
        return ResponseEntity.ok(user.get());
    }

    @DeleteMapping("/email/{email}")
    public ResponseEntity deleteUserByEmail(@PathVariable String email){
        errorResponse = new ErrorResponse(new Date());
        Optional<User> user = userService.deleteByEmail(email);
        if(user.isEmpty()){
            errorResponse.getErrorDescriptions().add("User with this email doesn't exist");
            return ResponseEntity.badRequest().body(errorResponse);
        }
        return ResponseEntity.ok(user.get());
    }

    @PutMapping("/id/{id}")
    public ResponseEntity updateUserById(@PathVariable Long id, @RequestBody User newUser){
        errorResponse = new ErrorResponse(new Date());
        Optional<User> user = userService.updateById(id, newUser);
        if(user.isEmpty()){
            errorResponse.getErrorDescriptions().add("User with this id doesn't exist");
            return ResponseEntity.badRequest().body(errorResponse);
        }
        return ResponseEntity.ok(user.get());
    }

    @PutMapping("/email/{email}")
    public ResponseEntity updateUserByEmail(@PathVariable String email, @RequestBody User newUser){
        errorResponse = new ErrorResponse(new Date());
        Optional<User> user = userService.updateByEmail(email, newUser);
        if(user.isEmpty()){
            errorResponse.getErrorDescriptions().add("User with this email doesn't exist");
            return ResponseEntity.badRequest().body(errorResponse);
        }
        return ResponseEntity.ok(user.get());
    }

    @GetMapping("/id/{id}")
    public ResponseEntity getUserById(@PathVariable Long id){
        errorResponse = new ErrorResponse(new Date());
        Optional<User> user = userService.findById(id);
        if(user.isEmpty()){
            errorResponse.getErrorDescriptions().add("User with this id doesn't exist");
            return ResponseEntity.badRequest().body(errorResponse);
        }
        return ResponseEntity.ok(user);
    }
    @GetMapping("/email/{email}")
    public ResponseEntity getUserByEmail(@PathVariable String email){
        errorResponse = new ErrorResponse(new Date());
        Optional<User> user = userService.findByEmail(email);
        if(user.isEmpty()){
            errorResponse.getErrorDescriptions().add("User with this email doesn't exist");
            return ResponseEntity.badRequest().body(errorResponse);
        }
        return ResponseEntity.ok(user);
    }
}

File: src/main/kotlin/cloud/folium/usercrud/presentation/UserController.kt
package cloud.folium.usercrud.presentation

import cloud.folium.usercrud.business.entity.User
import org.springframework.beans.factory.annotation.Autowired
import cloud.folium.usercrud.business.service.UserService
import kotlin.Throws
import javax.servlet.http.HttpServletRequest
import org.springframework.http.ResponseEntity
import org.springframework.validation.FieldError
import cloud.folium.usercrud.business.entity.UserRequestDto
import cloud.folium.usercrud.business.entity.annotations.Loggable
import org.springframework.http.HttpStatus
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.*
import java.io.IOException
import java.lang.Exception
import java.util.*
import javax.validation.Valid

@RestController
class UserController {
    @Autowired
    private lateinit var userService: UserService

    @ResponseStatus(HttpStatus.BAD_REQUEST)
    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleValidationExceptions(exception: MethodArgumentNotValidException): MutableMap<String, String?> {
        val errors: MutableMap<String, String?> = mutableMapOf()
        exception.bindingResult.allErrors.forEach{ error ->
            val fieldName: String = (error as FieldError).field
            val errorMessage: String? = error.defaultMessage
            if(!errors.containsKey(fieldName)) {
                errors[fieldName] = errorMessage
            } else {
                errors[fieldName] += " $errorMessage";
            }
        }
        return errors;
    }

    @PostMapping("/user")
    @Throws(Exception::class)
    fun addUser(
        @RequestBody @Valid user: User?,
        request: HttpServletRequest?
    ): ResponseEntity<*> {
        println("abcdefg")
        val newUser = userService.addUser(user, request)
        return if (newUser == null) {
            ResponseEntity.badRequest().body(
                Collections.singletonMap<String, Any>(
                    "error",
                    "User with email " + user!!.email + " already exist"
                )
            )
        } else ResponseEntity.ok(newUser)
    }

    @PostMapping(value = ["/user/multipart"])
    @Throws(IOException::class)
    fun addUser(
        @ModelAttribute userRequestDto: @Valid UserRequestDto?,
        request: HttpServletRequest?
    ): ResponseEntity<*> {
        val newUser = userService.addUser(userRequestDto, request)
        return if (newUser == null) {
            ResponseEntity.badRequest().body(
                Collections.singletonMap<String, Any>(
                    "error",
                    "User with email " + userRequestDto?.email + " already exist"
                )
            )
        } else ResponseEntity.ok(newUser)
    }

    @DeleteMapping("/user/id/{id}")
    fun deleteUserById(@PathVariable id: Long): ResponseEntity<*> {
        return if (userService.deleteById(id)) {
            ResponseEntity.ok(
                Collections.singletonMap(
                    "success",
                    "Deleted user with id $id"
                )
            )
        } else ResponseEntity.badRequest().body(
            Collections.singletonMap(
                "error",
                "User with id $id doesn't exist"
            )
        )
    }

    @GetMapping("/users/{country}")
    fun getAllUsersByCountry(@PathVariable country: String): ResponseEntity<*> {
        return ResponseEntity.ok(userService.getAllByCountry(country))
    }

    @GetMapping("/users")
    fun getAllUsers(): ResponseEntity<*> {
        return ResponseEntity.ok(userService.getAll())
    }

    @DeleteMapping("/user/email/{email}")
    fun deleteUserByEmail(@PathVariable email: String): ResponseEntity<*> {
        val deleted = userService.deleteByEmail(email)
        return if (deleted) {
            ResponseEntity.ok(
                Collections.singletonMap(
                    "success",
                    "Deleted user with email $email"
                )
            )
        } else ResponseEntity.badRequest().body(
            Collections.singletonMap(
                "error",
                "User with email $email doesn't exist"
            )
        )
    }

    @PutMapping("/user/id/{id}")
    fun updateUserById(@PathVariable id: Long, @RequestBody newUser: User): ResponseEntity<*> {
        val user = userService.updateById(id, newUser)
        return if (user == null) {
            ResponseEntity.badRequest().body(
                Collections.singletonMap(
                    "error",
                    "User with id $id doesn't exist"
                )
            )
        } else ResponseEntity.ok(user)
    }

    @PutMapping("/user/email/{email}")
    fun updateUserByEmail(@PathVariable email: String, @RequestBody newUser: User): ResponseEntity<*> {
        val user = userService.updateByEmail(email, newUser)
        return if (user == null) {
            ResponseEntity.badRequest().body(
                Collections.singletonMap(
                    "error",
                    "User with email $email doesn't exist"
                )
            )
        } else ResponseEntity.ok(user)
    }

    @GetMapping("/user/id/{id}")
    fun getUserById(@PathVariable id: Long): ResponseEntity<*> {
        val user = userService.findById(id)
        return if (user == null) {
            ResponseEntity.badRequest().body(
                Collections.singletonMap(
                    "error",
                    "User with id $id doesn't exist"
                )
            )
        } else ResponseEntity.ok(user)
    }

    @GetMapping("/user/email/{email}")
    fun getUserByEmail(@PathVariable email: String): ResponseEntity<*> {
        val user = userService.findByEmail(email)
        return if (user == null) {
            ResponseEntity.badRequest().body(
                Collections.singletonMap(
                    "error",
                    "User with email $email doesn't exist"
                )
            )
        } else ResponseEntity.ok(user)
    }
}